Scala 3 support

jtjeferreira opened the pull request:
Adds scala 3 support

lightbend-cla-validator commented:
Hi @jtjeferreira,

Thank you for your contribution! We really value the time you've taken to put this together.

We see that you have signed the Lightbend Contributors License Agreement before, however, the CLA has changed since you last signed it.
Please review the new CLA and sign it before we proceed with reviewing this pull request:

[http://www.lightbend.com/contribute/cla](http://www.lightbend.com/contribute/cla)

johanandren reviewed (commented):
We'll need a non-milestone Slick release before we can merge this to main. Good to have this prepared once that arrives though! 👍

jtjeferreira commented:
> Hey @jtjeferreira any chance you have time to rebase on master to solve the conflict so we can see if Slick 3.5.0 RC1 works?

I tried the "Resolve Conflicts" from the GitHub UI on my phone and the conflict seems simple. I will have a look tomorrow...

jtjeferreira commented:
> > Hey @jtjeferreira any chance you have time to rebase on master to solve the conflict so we can see if Slick 3.5.0 RC1 works?
> 
> I tried the "Resolve Conflicts" from the GitHub UI on my phone and the conflict seems simple. I will have a look tomorrow...

I fixed the conflicts. Can you approve the workflows? 

PS: There is a test that will fail, due to a compiler bug
```
[info] H2ScalaEventsByTagMigrationTest:
[info] - should migrate event tag to new way *** FAILED *** (2 milliseconds)
[info]   java.lang.ClassCastException: class com.typesafe.config.impl.SimpleConfig cannot be cast to class java.lang.String (com.typesafe.config.impl.SimpleConfig is in unnamed module of loader sbt.internal.LayeredClassLoader @611cd810; java.lang.String is in module java.base of loader 'bootstrap')
[info]   at akka.persistence.jdbc.query.EventsByTagMigrationTest.config$accessor(EventsByTagMigrationTest.scala:31)
[info]   at akka.persistence.jdbc.query.EventsByTagMigrationTest.withRollingUpdateActorSystem(EventsByTagMigrationTest.scala:161)
[info]   at akka.persistence.jdbc.query.EventsByTagMigrationTest.testFun$proxy1$1(EventsByTagMigrationTest.scala:188)
[info]   at akka.persistence.jdbc.query.EventsByTagMigrationTest.$init$$$anonfun$1(EventsByTagMigrationTest.scala:171)
```

I will push the fix after so we can have proof of the bug (and the bugfix)

jtjeferreira commented:
Thanks for approving the workflows. There are 2 issues:
* I haven't changed the workflows to use scala3. Any tip where I should do that? `.github/workflows/checks.yml`? Any examples from other akka projects I could follow? 
* Mima warnings in 2.13 on `JdbcBackend#DatabaseDef` vs `JdbcBackend#JdbcDatabaseDef` which [changed](https://github.com/slick/slick/pull/2529) in slick 3.5.0. Can I add a mima filter?

johanandren commented:
I pushed a commit adding Test/compile on Scala 3 to CI PR validation now. There are some related failures to look into - methods that end up with the same signature after erasure.

johanandren commented on core/src/main/mima-filters/5.4.0.backwards.excludes/issue-775-slick-3.50.excludes:
```diff
@@ -0,0 +1,38 @@
+ProblemFilters.exclude[IncompatibleResultTypeProblem]("akka.persistence.jdbc.db.EagerSlickDatabase.copy$default$1")
+ProblemFilters.exclude[IncompatibleMethTypeProblem]("akka.persistence.jdbc.db.EagerSlickDatabase.this")
+
+ProblemFilters.exclude[IncompatibleResultTypeProblem]("akka.persistence.jdbc.db.LazySlickDatabase.database")
```
I think these should be internal API anyway, please mark both `EagerSlickDatabase` and `LazySlickDatabase` as such.

johanandren commented on core/src/main/mima-filters/5.4.0.backwards.excludes/issue-775-slick-3.50.excludes:
```diff
@@ -0,0 +1,38 @@
+
+ProblemFilters.exclude[IncompatibleResultTypeProblem]("akka.persistence.jdbc.db.SlickDatabase.forConfig")
+ProblemFilters.exclude[IncompatibleResultTypeProblem]("akka.persistence.jdbc.db.SlickDatabase.database")
+ProblemFilters.exclude[ReversedMissingMethodProblem]("akka.persistence.jdbc.db.SlickDatabase.database")
```
`SlickDatabase` already is internal, so this is fine

johanandren commented on core/src/main/mima-filters/5.4.0.backwards.excludes/issue-775-slick-3.50.excludes:
```diff
@@ -0,0 +1,38 @@
+ProblemFilters.exclude[IncompatibleMethTypeProblem]("akka.persistence.jdbc.snapshot.dao.DefaultSnapshotDao.this")
+ProblemFilters.exclude[IncompatibleMethTypeProblem]("akka.persistence.jdbc.snapshot.dao.legacy.ByteArraySnapshotDao.this")
+ProblemFilters.exclude[IncompatibleResultTypeProblem]("akka.persistence.jdbc.state.JdbcDurableStateStoreProvider.db")
+ProblemFilters.exclude[IncompatibleMethTypeProblem]("akka.persistence.jdbc.state.scaladsl.JdbcDurableStateStore.this")
```
Both fine to exclude:

Constructor isn't used by user code (instantiated by akka-persistence). A bit curious that it shows up, I can't see that it wasn't touched for any of the listed types though.

`db` field shouldn't ever be touched by user code, effectively protected.

johanandren commented on core/src/main/scala/akka/persistence/jdbc/query/JdbcReadJournalProvider.scala:
```diff
@@ -11,7 +11,8 @@ import com.typesafe.config.Config
+    new scaladsl.JdbcReadJournal(config, configPath)(system)
 
-  override val javadslReadJournal = new javadsl.JdbcReadJournal(scaladslReadJournal)
+  override def javadslReadJournal(): javadsl.JdbcReadJournal = new javadsl.JdbcReadJournal(scaladslReadJournal())
```
The returned instances are cached and shared by akka-persistence internals so this change is fine. 👍

johanandren commented on core/src/main/scala/akka/persistence/jdbc/query/JournalSequenceActor.scala:
```diff
@@ -38,15 +38,15 @@ object JournalSequenceActor {
     def isEmpty: Boolean = elements.forall(_.isEmpty)
   }
-  private object MissingElements {
+  object MissingElements {
```
Make the multi-param `JournalSequenceActor.receive(currentMaxOrdering ...)` private instead of changing these (it's an internal factory for `Receive ` instances, not called from the outside)

johanandren commented on core/src/main/scala/akka/persistence/jdbc/state/scaladsl/DurableStateSequenceActor.scala:
```diff
@@ -46,7 +46,7 @@ import akka.annotation.InternalApi
    * It can be seen as a collection of GlobalOffset
    */
-  private case class MissingElements(elements: Seq[NumericRange[GlobalOffset]]) {
+  case class MissingElements(elements: Seq[NumericRange[GlobalOffset]]) {
```
Same here, make the method private instead

johanandren reviewed (commented):
Looks good so far! Some smaller things and the conflicting overloads and this is as ready as it gets until slick 3.5.0 is released 👍

johanandren commented:
> If I comment the @ApiMayChange it compiles. If I add it again and compile incrementally it works... I guess this is another compiler bug?

Sounds like that, removing that annotation and adding Scaladoc saying it "API may change" is fine IMO

jtjeferreira commented:
> > If I comment the @ApiMayChange it compiles. If I add it again and compile incrementally it works... I guess this is another compiler bug?
> 
> Sounds like that, removing that annotation and adding Scaladoc saying it "API may change" is fine IMO

Meanwhile, I found a workaround (see 45bbe4a).

I think I addressed all your feedback and also fixed mima after e6422a4

johanandren commented:
Even better 👍

jtjeferreira commented:
Regarding the failure in `H2ScalaCurrentEventsByTagTest#should complete without any gaps in case events are being persisted when the query is executed` I was also facing the problem locally, but I was hoping it was due to some slowness in my machine. 

I fixed it with 7199f96, but I wonder if this is a problem with slick 3.5.0...

johanandren reviewed (approved):
LGTM, thanks for the good work @jtjeferreira !

jtjeferreira commented:
> LGTM, thanks for the good work @jtjeferreira !

Thanks. Looking forward for a release

jtjeferreira commented:
Looks like there is a problem in docs https://github.com/akka/akka-persistence-jdbc/actions/runs/8201918926/job/22431619062:

```
[error] missing argument for option -skip-packages
```
